### ISSUE TITLE
[css-overflow] Line-clamp and screen readers

### ISSUE THREAD

**Issue by frivoal** [w3c/csswg-drafts#12859]
What should happen to content clamped by `line-clamp` in the accessibility tree?

Just like content made invisible to sighted users by `visibility: hidden`, I think that content made invisible to sighted users using `line-clamp` should also be hidden from screen-reader users.

**Comment by SebastianZ** [comment]
The same question probably also applies to `overflow`, maybe in combination with `text-overflow`.

> I think that content made invisible to sighted users using `line-clamp` should also be hidden from screen-reader users.

I am not sure about that. Line clamping is often used to harmonize layout, e.g. when you have several article teasers on a page or you have a list/table with titles in it. In those cases authors often limit the number of lines or the width of the column to avoid "fringing", i.e. visually cluttering the page. Though often that effect is only meant to be visual and authors sometimes display the full text in a tooltip.
In such cases, I'd expect a screen reader to read the full text.

On the other hand, if an author clamps a full text that would otherwise span many lines to just two or three, it might make sense to also hide that from screen readers. The question then is, what they are supposed to output in those cases. Just stopping in the middle of a sentence or even a word seems pretty bad for UX.

Sebastian

**Comment by frivoal** [comment]
> The same question probably also applies to overflow

I'm not convinced. `overflow: scroll` hides content, but also gives you a way to access it. `line-clamp` just hides (and if a way to access the hidden content is provided by the author, it's by removing `line-clamp`).

If the full text is displayed somehow (including in a tooltip), then I'd think about it differently, but as specified, that is not the case, the remainder of the content is just not visible.

In general, I think screen readers, as their names indicates, are intended to give an audio rendition of what's on screen, not an independent audio rendition of the source document regardless of how it's styled.

If something is invisible yet read aloud, I believe that can be confusing, for instance in cases where a (partially) sighted user is accessing the page with a screen reader, and cannot easily reconcile what they're hearing with what they're seeing, or in case where a blind user discusses the content of the page with a sighted user, and they have a different understanding of what's there.

> Just stopping in the middle of a sentence or even a word seems pretty bad for UX.

That is exactly the UX that a sighted user is getting, though. As long as the ellipsis or some audio equivalent is rendered, I'm not certain that's an issue.

That said, I am not personally a screen reader user, nor have I don't user research on this specific question, so my guesses as to what's best are just that: guesses.

**Comment by frivoal** [comment]
> The same question probably also applies to overflow

I'm not convinced. `overflow: scroll` hides content, but also gives you a way to access it. `line-clamp` just hides (and if a way to access the hidden content is provided by the author, it's by removing `line-clamp`)

**Comment by SebastianZ** [comment]
> > The same question probably also applies to overflow
> 
> I'm not convinced. `overflow: scroll` hides content, but also gives you a way to access it. `line-clamp` just hides (and if a way to access the hidden content is provided by the author, it's by removing `line-clamp`).

Yes, `overflow: scroll` allows users to access the text, though `hidden` and `clip` don't. (Well, `hidden` does programmatically.) And those two values are usually used when authors work with `text-overflow`.

> In general, I think screen readers, as their names indicates, are intended to give an audio rendition of what's on screen, not an independent audio rendition of the source document regardless of how it's styled.

In general yes, with some exceptions like decorative styling. And I'd argue that truncating text may fall into that category, at least in some cases.

And https://butterpep.com/line-clamp-overflow-ellipsis.html indicates that screen readers do read the full text (but the tested versions are pretty outdated).

> If something is invisible yet read aloud, I believe that can be confusing, for instance in cases where a (partially) sighted user is accessing the page with a screen reader, and cannot easily reconcile what they're hearing with what they're seeing, or in case where a blind user discusses the content of the page with a sighted user, and they have a different understanding of what's there.

I'd say, any confusion in those cases only depends on how much of the content is truncated.

> > Just stopping in the middle of a sentence or even a word seems pretty bad for UX.
> 
> That is exactly the UX that a sighted user is getting, though. As long as the ellipsis or some audio equivalent is rendered, I'm not certain that's an issue.

Screen readers usually don't interpret the context. So, a cut-off word may be read differently than the full word, which may cause confusion.

> That said, I am not personally a screen reader user, nor have I don't user research on this specific question, so my guesses as to what's best are just that: guesses.

I couldn't find a proper rule in the WCAG explicitly targetting truncated text. The only one that comes near seems to be https://www.w3.org/WAI/WCAG22/Understanding/resize-text.

And as we both are sighted users, I guess the best would be if people using screen readers and/or the Accessibility Guidelines Working Group weighed in here.

Sebastian

**Comment by andreubotella** [comment]
Discussing this with some folks at Igalia, it seems like if the clamped content can't be selected and copied, then it also shouldn't be exposed to screen readers. And the current spec says that the boxes and line boxes after clamp become [invisible boxes](https://drafts.csswg.org/css-display-4/#invisible-box), which indeed are unselectable (although I hadn't taken that into account in my Chromium implementation, and we should write WPTs for that).

**Comment by cookiecrook** [comment]
To clarify, there is not really an option to “hide from screen readers” but show to other AT. It’s either removed from the API, or not.

So I mainly worry about hiding this, in the context of say: Larger Text sizes, in combination with VoiceOver or HoverText (both of these are common uses)

Fonts at 200% plus a line clamp may become unreasonably short, unreadable text strings. A sighted VO or HoverText user would access these through the API. If these are hidden, they are out of luck. 

I would say that if there is ever a possibility that the text may be rendered (for example, small text on a large wide desktop display) then the full text should be exposed to the accessibility API.

**Comment by frivoal** [comment]
> I would say that if there is ever a possibility that the text may be rendered (for example, small text on a large wide desktop display) then the full text should be exposed to the accessibility API.

If the text may be rendered, that's only when the layout changes, at which point it is no longer clamped, and would be exposed to accessibility APIs.

Am I missing something?

**Comment by cookiecrook** [comment]
Yes. 

Let’s say we have the following sentence:

“Lorem ipsum dolor sit amet, consectetur adipiscing elit.”

And in the default font size the line clamp renders:

> Lorem ipsum dolor sit 
> amet, consectetur…

And in the case of a low-vision, sighted screenreader user (who uses a combination of Larger Text, VoiceOver, Zoom, and HoverText) the font might be extra large, rendering only the following:

> Lorem 
> ipsum…

But a different sighted user with better than average vision (or a visually impaired person with high clarity but low peripheral vision) could also set their font size to xx-small, and the whole sentence would be rendered:

> Lorem ipsum dolor sit amet, 
> consectetur adipiscing elit. 

In the above three possible renderings of this single hypothetical scenario, the low vision screen reader user should not be penalized for displaying large text, especially if the large text benefits their remaining usable vision. Clamping the Accessibility API label value to the visibly rendered text would prevent their screen reader from voicing the entire text, and it would prevent tools like HoverText from being able to re-render the text in a large tooltip-like view, the same text that their sighted colleagues can read visually.

Moving on from the low vision screen reader user example, many fully blind screen reader users don’t always know (nor should they have to care) whether their browser window is rendered large or small. Changing what is exposed to the accessibility tree this way could leave them with unpredictable results. One day being able to access content at the end of the line clamp, and the next day finding that content missing due to it being rendered in a slightly smaller window.

**Comment by SebastianZ** [comment]
Given the use cases above, I wonder whether it would make sense to expose both to the accessibility tree, the whole content _and_ the information how it is rendered. This allows the screen reader ‒ and in extension the user if the screen reader provides the option for it ‒ to decide, whether hidden text is read or not.

Sebastian

**Comment by frivoal** [comment]
thanks @cookiecrook , that did help…

However, I am now differently confused: When people resize the font size or do that kind of change, various css features, notably mediaqueries, might completely change the layout of the page. I thought a general principle of screen readers and other accessibility tools was to take into account the actual layout to present information that corresponds to what is being seen. If we go after information that corresponds to what might be seen if zoom levels / font sizes were different, that's a different target altogether, and one which can vary so arbitrarily that it makes me wonder if layout should be taken into account at all, instead simply working from the DOM and nothing else. But that's clearly not what we're doing in general…

Another concern I have here is that while reading the whole text if we're showing 3 lines out of 4 or 5 seems ok, it might also be 3 lines out of 500. For example, on a page where a bunch of short excerpts of long articles are shown, and you can click to expand which one you want to read. If we render the whole articles instead of the short excerpts to the accessibility tree, we've changed an overview and navigation page into a long-form multi-chapter document.

**Comment by frivoal** [comment]
@cookiecrook, do you think Sebastian's approach from https://github.com/w3c/csswg-drafts/issues/12859#issuecomment-4064818600 be practical? 

I don't know if there's precedent for showing two different "views" of the same element in the accessibility tree. If there is, this would seem promising, but if not, that might be too much of a new concept to introduce.

**Comment by emilio** [comment]
I think this is also pretty weird behavior. I guess we should also make stuff that's clamped unfocusable? Or not?

**Comment by frivoal** [comment]
> I think this is also pretty weird behavior

Can you clarify which behavior you consider to be pretty weird?

> I guess we should also make stuff that's clamped unfocusable?

That would be my expectation. Having invisible content be focusable would be pretty confusing.

**Comment by SebastianZ** [comment]
> Another concern I have here is that while reading the whole text if we're showing 3 lines out of 4 or 5 seems ok, it might also be 3 lines out of 500. For example, on a page where a bunch of short excerpts of long articles are shown, and you can click to expand which one you want to read. If we render the whole articles instead of the short excerpts to the accessibility tree, we've changed an overview and navigation page into a long-form multi-chapter document.

That was my initial point. Exposing the full contents in those cases is as bad as exposing only the visible part (without any indication that there's more content).

One might argue, this use case could be covered by using `<details>` and `<summary>` elements, which only expose their contents once they are opened. Though the summary would then have to duplicate the contents and actually be ellipsized in HTML. So that's not the same.

So I think either UAs have to be smart about what to expose to the AT in that case or authors need at least _some_ control over what gets exposed.

It reminds me a little of #560 and #9579. Though in this case we are not in an all-or-nothing situation but authors might rather want to expose only part of the actual contents.

So my idea is to introduce a new property that controls whether the whole content is exposed to AT or only the unclamped parts with an indication that there's more. And UAs should have a sensible default value for that.

Sebastian

**Comment by jcsteh** [comment]
As is often the case, the correct behaviour here comes down to the author's intent. If the author ultimately intends all of the text to be perceivable but the text can't fit purely for reasons of visual layout, then I'd argue the text should be exposed to accessibility APIs. There are a lot of truncation cases like this. The visual parallel might be that the text is exposed with a tooltip, that you can zoom to expose more of the text, etc. Personally, I lean more in this direction in this case.

On the other hand, if the author does not intend all of the text to be exposed - i.e. this is not purely a presentational restriction, but is rather a fundamental intention of the UI - then it shouldn't be exposed to accessibility APIs.

This is another situation where I think we hit a conflict of principles. CSS is generally thought of as being specific to visual styling, so the general sense is that CSS properties are less about intent and more about visual layout. Thus, there's a tendency to define CSS properties purely with respect to their visual impact, covering as many use cases as possible, and therefore the answer is "we don't know what the author intends". But that conflicts somewhat with accessibility, where the intent/use case is fundamental to choosing the correct behaviour. Ideally, I think the property (or properties) need to make it clear what the intent is, rather than this being a guessing game on the part of authors, accessibility APIs, assistive tech or users.

**Comment by SebastianZ** [comment]
Very well expressed, @jcsteh! Two points I'd add to this:

If it's the authors intent to _not_ expose all of the text - neither visually nor aurally - then the cropping should already happen in the DOM. If the cropping is meant to only be presentational, then I'd say this should be done in CSS, visually _and_ aurally.

And just to emphasize this again, CSS is not just about _visual_ rendering of contents but any form of presentation, hence the [CSS Speech spec](https://drafts.csswg.org/css-speech-1/), [CSS Spatial Nav](https://drafts.csswg.org/css-spatial-nav-1/), etc. So again, authors should get at least _some_ control over what gets exposed to AT to be able to express their intent, but it is also important to provide a good default (like it's always the case for any CSS feature we introduce).

Sebastian

**Comment by johannesodland** [comment]
A fairly common design requirement we get is to show a few rows of a table, a few items of a list or a few lines of text, and allow the user to click a button to expand and see the rest.

The intent here is to show the most important parts, or tease a much longer content, and allow the most interested users to expand for a deep dive, while allowing "regular" users to skip past. This is often used with result tables, or editorial deep dives like "This is the method we used to find...". 

With tables and lists, this is easy to implement by hiding elements with `display:none`, but with free form text, it is more complicated as there are no "line elements" to hide. `line-clamp` will help in that regards.

Since the intent is to hide the remaining content from all users, it should be possible to hide it from AT as well. Giving them a different experience would be confusing. Additionally, we could end up trapping them inside a long deep dive that was intended to be hidden.

If the clamped content is exposed to AT by default, we would need a new property that allows authors to hide it if that was the intention.

**Comment by dshin-moz** [comment]
This is marked resolved, but has since has had sizeable discussion that is worth revisiting. Would be interested since we're working on `line-clamp` for Gecko as well.

**Comment by thejohnjansen** [comment]
After discussion with the ARIA working group, it was resolved that we are behaving as expected. There are beneficial reasons for doing so (e.g., when combined with zoom, the text may become almost invisible but is still read aloud). I have submitted a PR with 2 tests that validate this behavior. https://chromium-review.googlesource.com/c/chromium/src/+/8352688

I think this issue can be closed @frivoal

**Comment by thejohnjansen** [comment]
@flackr , can I add you as a reviewer for the two simple tests I wrote validating this behavior?
